File: data-structures/binary-tree/binaryTree.js
const binaryTree = () => {
    let tree = null;

    const insertInternal = (value, node) => {
        if (node === null) {
            return { value, left: null, right: null };
        }

        if (value < node.value) {
            node.left = insertInternal(value, node.left);
        } else if (value >= node.value) {
            node.right = insertInternal(value, node.right);
        }

        return node;
    };

    const insert = value => {
        tree = insertInternal(value, tree);
    };

    const contain = (value, node = tree) => {
        if (node === null) return false;
        if (value === node.value) return true;
        if (value < node.value) return contain(value, node.left);
        if (value > node.value) return contain(value, node.right);
    };

    return {
        insert,
        contain,
    };
};

const { insert, contain } = binaryTree();

for (const elem of [
    99, 6, 81, 52, 58, 23, 4, 3, 90, 7, 12, 69, 20, 42, 31, 34, 68, 61, 41, 55,
    4, 3, 7, 93, 14, 76, 27,
]) {
    insert(elem);
}

console.log(contain(3));
console.log(contain(4));
console.log(contain(93));
console.log(contain(2));

File: data-structures/binaryTree/binaryTree.go
package main

import "fmt"

type IntBinaryTree struct {
	val   int
	left  *IntBinaryTree
	right *IntBinaryTree
}

func (bt *IntBinaryTree) Insert(value int) *IntBinaryTree {
	if bt == nil {
		return &IntBinaryTree{val: value}
	}

	if value < bt.val {
		bt.left = bt.left.Insert(value)
	} else if value >= bt.val {
		bt.right = bt.right.Insert(value)
	}

	return bt
}

func (bt *IntBinaryTree) Contains(value int) bool {
	switch {
	case bt == nil:
		return false
	case value < bt.val:
		return bt.left.Contains(value)
	case value > bt.val:
		return bt.right.Contains(value)
	default:
		return true
	}
}

func main() {
	var bt *IntBinaryTree
	bt = bt.Insert(5)
	bt = bt.Insert(3)
	bt = bt.Insert(10)
	bt = bt.Insert(2)

	fmt.Println(bt.Contains(2))
	fmt.Println(bt.Contains(3))
	fmt.Println(bt.Contains(11))
	fmt.Println(bt.Contains(10))
}
